src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperComponentExtensions.cs

```diff
@@ -46,39 +46,28 @@ public static class HtmlHelperComponentExtensions
         {
             if (htmlHelper == null)
             {
-                throw new System.ArgumentNullException(nameof(htmlHelper));
+                throw new ArgumentNullException(nameof(htmlHelper));
             }
 
-            var serviceProvider = htmlHelper.ViewContext.HttpContext.RequestServices;
-            var encoder = serviceProvider.GetRequiredService<HtmlEncoder>();
+            var httpContext = htmlHelper.ViewContext.HttpContext;
+            var serviceProvider = httpContext.RequestServices;
+            var prerrenderContext = serviceProvider.GetRequiredService<MvcPrerrenderingContext>();
+            var encoder = prerrenderContext.Encoder;
+            prerrenderContext.Environment.Name = ComponentEnvironment.Prerrender;
+            // Add our custom JSRuntime that throws upon invocation. We might need an option here to not throw or log a warning
+            // message instead.
```

Comment:
At least for now, I'd be strongly in favour of throwing like you have here. Silently ignoring JS interop calls would almost always produce unhelpful behavior. People who are doing JS interop in combination with prerendering need to think through how they want to change their application behavior/output in the prerendering case, and the exception forces them to do that.